fixieai/agents/agent_func.py

```diff
@@ -0,0 +1,226 @@
+from __future__ import annotations
+
+import collections.abc
+import inspect
+import typing
+from typing import Any, Callable, Iterable, Optional, Tuple, Union, get_type_hints
+
+from fixieai.agents import api
+from fixieai.agents import oauth
+from fixieai.agents import user_storage
+
+# An ArgumentMapper converts the a query+agent ID into a supported func argument (e.g. "query" or "user_storage")
+ArgumentMapper = Callable[[api.AgentQuery, str], Any]
+BoundArgumentMapper = Tuple[str, ArgumentMapper]
+
+
+class AgentFunc:
+    def __init__(
+        self,
+        impl: Callable,
+        argument_mappers: Iterable[BoundArgumentMapper],
+        allow_multiple_responses: bool,
+    ):
+        self._impl = impl
+        self._argument_mappers = tuple(argument_mappers)
+        self._allow_multiple_responses = allow_multiple_responses
+
+    def __call__(
+        self, agent_query: api.AgentQuery, agent_id: str
+    ) -> Iterable[api.AgentResponse]:
+        kwargs = {}
+        for name, mapper in self._argument_mappers:
+            kwargs[name] = mapper(agent_query, agent_id)
+
+        result = self._impl(**kwargs)
+        return _adapt_func_result(result, self._allow_multiple_responses)
+
+    @classmethod
+    def create(
+        cls,
+        func: Callable,
+        oauth_params: Optional[oauth.OAuthParams],
+        default_message_type: type,
+        allow_generator: bool,
+    ) -> AgentFunc:
+        if not inspect.isfunction(func):
+            raise TypeError(
+                f"Registered function {func!r} is not a function, but a {type(func)!r}."
+            )
+        signature = inspect.signature(func)
+        func_name = func.__name__
+        params = signature.parameters
+
+        # Validate that there are no var args (*args or **kwargs).
+        if any(
+            param.kind in (param.VAR_KEYWORD, param.VAR_POSITIONAL)
+            for param in params.values()
+        ):
+            raise TypeError(
+                f"Registered function {func_name} cannot accept variable args: {params!r}."
+            )
+
+        # Ensure every parameter is used exactly once.
+        unresolved_parameters = set(name for name in params.keys())
+
+        # Bind a query/message mapper.
```

Comment:
Please write a comment paragraph explaining what's going to happen in the rest of this method! It takes a bit time to understand it, especially to someone without context.